### ISSUE TITLE
Behaviour change: Change ruby style from "prefer single quotes" to "always use double quotes"

### ISSUE THREAD

**Issue by DavidS** [puppetlabs/pdk-planning#17]
From the early days of using rubocop I have inherited the default style of "prefer single quotes when possible". As it turns out, this is a unhelpful style as it adds additional cognitive burden to developers (especially the majority of our target user group, who are not full-time ruby developers) in having to understand two sets of quoting rules, and when to switch up from one to the other when modifying code.

To avoid that burden, I propose switching to a "always use double quotes" style. This might require a bit more backslash-escaping when there is no "technical" need for it (because the string doesn't contain a variable reference, and could be simplified by using single quotes), but - as referenced in the first paragraph - it'll reduce the amount of detailed ruby knowledge required overall.

### References
* https://github.com/rubocop-hq/rubocop/issues/5306
* @ripienaar on slack: "`'` vs `"` is 100% the hill I will die on :slightly_smiling_face:"
* ruby defaults to double quotes on output: `ruby -e "puts 'foo'.inspect"` :arrow_right: `"foo"`

**Comment by scotje** [comment]
👍  Double quotes everywhere is my preference as well. My understanding is that the performance implications for most use cases are rounding errors at this point.

![](https://i.imgflip.com/2f3lcp.jpg)

**Comment by scotje** [comment]
@jbondpdx What would be the process to get the Language Style Guide updated to reflect the quoting preference change proposed here?

**Comment by jbondpdx** [comment]
@DavidS and @scotje This was discussed ad nauseum in 2015-2016, and at the modules team gathering in Belfast in 2016(?) we decided to leave the recommendation (and style guide) alone, at least in part due to the amount of work it would take to change this in existing code. You can enjoy the JIRA comments here: https://tickets.puppetlabs.com/browse/DOCUMENT-274

Fortunately, we didn't etch it in stone or anything! Traditionally, modules has "owned" style guide, so I recently handed the style guide off to @clairecadman . Of course, in addition to modules team, our team has a vested interest in the style guide, as do Language (Henrik and Thomas) and the community.

I think the first move should be a style guide ticket (DOCUMENT project) assigned to Claire. It's my personal opinion that changing code style is kind of a big deal, and changes should be decided by a cross-team group representing the aforementioned stakeholders. That said, it's Claire's to drive!

**Comment by clairecadman** [comment]
@DavidS @scotje Yes please create a docs ticket and assign it to me if you want updates to the style guide. Thanks!

**Comment by GeoffWilliams** [comment]
I wrote the original email that Jean captured in that ticket. I still fully stand by my original comment and 100% approve of this change ticket.
I already disable this lint check on all of my puppet code which also gets the "quality" scoring of my forge modules marked down.
I encourage Puppet to do everything possible to reduce the cognitive burden on our users.

**Comment by jorhett** [comment]
I completely disagree with this approach. I fix far too many problems created by people not realizing what character combinations do, which ones are meta characters, etc. Forcing everyone to use double quotes all the time demands that everyone be aware of all meta combinations, all interpolation rules, etc etc. Talk about cognitive load!

That said, I do understand that some people feel otherwise. I'd much prefer that we allow either pattern in puppet-lint, and don't mark down either one used consistently.

**Comment by GeoffWilliams** [comment]
So in my original email back in 2015 I said "Use double quotes unless you have a good reason not to" When my students struggled with windows backslashes and strings with double quotes themselves, I didn't stand over them with a big stick and force them to figure it out - I just said "well that's a good reason" and told them do whatever was easiest.

My whole rationale behind this is that arguing over what type of quotes to use is a zero-sum game. It's a complete waste of productive energy that could have been used to put things _inside_ those quotes - double or otherwise.

**Comment by scotje** [comment]
Perhaps the style guide recommendation should be something more along the lines of "pick a default style for quoting string literals and apply it consistently to your codebase".

PDK will still end up having to have a default for new modules but we should also leverage the [config subsystem](https://github.com/puppetlabs/pdk-planning/blob/master/RFCs/0002-add-pdk-config.md) to allow users to set their preferred quoting style system-wide, and have PDK read from the config value when applying the default template.

Thoughts?

**Comment by jorhett** [comment]
I would love that. Allow each side to configure it for their own modules as they see fit, and thus allow advocates for both sides to coexist without conflict.

**Comment by scotje** [comment]
Opened https://tickets.puppetlabs.com/browse/DOCUMENT-931 to discuss changes to the style guide as requested.

**Comment by DavidS** [comment]
Given the strong and divergent opinions around this topic, I support @scotje's suggestion to step away from recommending a specific quoting style, and only provide a `ignore`/`force to single`/`force to double` toggle in `.sync.yml` so that a module author can choose and enforce their preferred style.

**Comment by scotje** [comment]
Here is a draft of the update to the quoting section of the style guide, please comment on the Google doc if you have any feedback!

https://docs.google.com/document/d/1oIzh4FAzwVOn7BIWEFDAGxHnucaxv0I18J9j4tsVLwo/edit?usp=sharing

**Comment by ardrigh** [comment]
I just want to comment my support for the _Let the people decide_ option to score it being _consistent_ in the codebase, but not enforced to be one over the other.

A problem for one person is often only a corner-case for someone else, and thus users should be able to decide.

Our internal style guide is to _use single quotes except in cases where interpolation is explicitly required_.   I find this works well with modules and Hiera YAML.

**Comment by seanmil** [comment]
I'm usually all in favor of "let the people decide", but now you have shifted the cognitive burden significantly to contributors of a variety of modules/projects.

If, as a community, we are all using the same quoting rules then you must learn those rules once and learn how to apply them once. That burden may be high, but now if you let each group/developer pick their own not only do you need to learn those rules you now also have to learn what the rules are for a particular project and incur the mental overhead of having to switch during the transition between different projects with different rules and fight against muscle memory when typing.

I understand the desire to lower the cognitive burden on someone new to Puppet, and if PDK wants to default to "ignore quotes" to ease that developer's introduction to Puppet, especially for that developer's own/private modules, then I think that would be okay. In the interest of choice, I think even letting force-single / force-double to be okay options as well.

But please please please do not change the style guide or the Forge checks. There is a huge advantage in having the community and community contributions consistent as a whole. The amount of uniformity in code style lowers the bar significantly for first time contributors (and people learning from reading others' code).

**Comment by scotje** [comment]
@seanmil I think the cognitive burden side of varying preferred quoting styles can and should be resolved by automatic tooling. (e.g. `rubocop -a`) and if there are gaps in the ability for tooling to correct things we should focus our efforts on fixing those gaps.

There is always going to be some degree of cognitive burden associated with switching between various projects and we are still setting very specific guidelines about how to apply each of the two styles so the variation will be pretty narrow in scope.

**Comment by bmjen** [comment]
Style guide was updated to address single vs double quote preferences in Puppet Code.
https://puppet.com/docs/puppet/6.0/style_guide.html#quoting

**Comment by scotje** [comment]
Remaining changes here would be to update puppet-lint checks/defaults as needed and potentially add a default to PDK's rubocop config. @rodjek what do you think the best course of action is for puppet-lint?
